File: src/ycp/scoring.py
"""Scoring engine — the deterministic core of the closed loop.

Turns per-clip metrics into a 0–100 virality score, then rolls up by dimension
(source creator, format, hook, length, platform) so we can see which *combos*
over-index. Pure pandas: fully reproducible and unit-tested. No LLM here — the
LLM only writes prose in brief.py; the numbers are always deterministic.
"""
from __future__ import annotations

import numpy as np
import pandas as pd

from .config import settings

ACTION_DIMS = ["source_creator", "fmt", "hook_type"]  # the actionable combo


def _minmax(s: pd.Series) -> pd.Series:
    """Scale to 0..1. Constant or empty series -> 0.5 (neutral, no signal)."""
    if s.empty:
        return s
    lo, hi = s.min(), s.max()
    if not np.isfinite(lo) or not np.isfinite(hi) or hi == lo:
        return pd.Series(0.5, index=s.index)
    return (s - lo) / (hi - lo)


def length_bucket(seconds: float | int | None) -> str:
    buckets = settings()["scoring"]["length_buckets"]
    if seconds is None or (isinstance(seconds, float) and np.isnan(seconds)):
        return "unknown"
    for lo, hi in zip(buckets, buckets[1:]):
        if lo <= seconds < hi:
            return f"{lo}-{hi}s"
    return f"{buckets[-1]}s+"


def add_derived(df: pd.DataFrame) -> pd.DataFrame:
    """Add revenue_per_1k and length_bucket. Returns a new frame (immutable in)."""
    if df.empty:
        return df.copy()
    out = df.copy()
    views = out["views"].clip(lower=0).fillna(0)
    revenue = out.get("ad_revenue", 0).fillna(0) + out.get("whop_payout", 0).fillna(0)
    with np.errstate(divide="ignore", invalid="ignore"):
        out["revenue_per_1k"] = np.where(views > 0, revenue / (views / 1000.0), 0.0)
    out["retention_pct"] = out.get("retention_pct", pd.Series(index=out.index)).fillna(0)
    out["length_bucket"] = out["length_sec"].apply(length_bucket)
    return out


def compute_scores(df: pd.DataFrame, weights: dict | None = None) -> pd.DataFrame:
    """Add a 0–100 `virality_score` blending views, retention, revenue/1k."""
    if df.empty:
        return df.copy()
    w = weights or settings()["scoring"]["weights"]
    out = add_derived(df)
    n_views = _minmax(np.log1p(out["views"].clip(lower=0).fillna(0)))
    n_ret = _minmax(out["retention_pct"].clip(lower=0))
    n_rev = _minmax(out["revenue_per_1k"].clip(lower=0))
    score = (
        w["views_7d"] * n_views
        + w["retention"] * n_ret
        + w["revenue_per_1k"] * n_rev
    )
    out["virality_score"] = (score * 100).round(1)
    return out


def rollup(df: pd.DataFrame, dims: list[str], min_sample: int | None = None) -> pd.DataFrame:
    """Group scored clips by `dims`; return mean score/views, total revenue, n.

    Combos with fewer than `min_sample` clips are dropped (not enough signal).
    """
    if df.empty or "virality_score" not in df:
        return pd.DataFrame(columns=[*dims, "n", "avg_score", "avg_views", "total_revenue"])
    min_sample = settings()["scoring"]["min_sample"] if min_sample is None else min_sample
    df = df.copy()
    df["_revenue"] = df.get("ad_revenue", 0).fillna(0) + df.get("whop_payout", 0).fillna(0)
    g = (
        df.groupby(dims, dropna=False)
        .agg(
            n=("clip_id", "count"),
            avg_score=("virality_score", "mean"),
            avg_views=("views", "mean"),
            total_revenue=("_revenue", "sum"),
        )
        .reset_index()
    )
    g = g[g["n"] >= min_sample]
    g["avg_score"] = g["avg_score"].round(1)
    g["avg_views"] = g["avg_views"].round(0)
    g["total_revenue"] = g["total_revenue"].round(2)
    return g.sort_values("avg_score", ascending=False).reset_index(drop=True)


def scale_and_kill(
    rolled: pd.DataFrame, scale_q: float | None = None, kill_q: float | None = None
) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Split a rollup into top-quantile (scale) and bottom-quantile (kill)."""
    cfg = settings()["scoring"]
    scale_q = cfg["scale_quantile"] if scale_q is None else scale_q
    kill_q = cfg["kill_quantile"] if kill_q is None else kill_q
    if rolled.empty:
        return rolled, rolled
    hi = rolled["avg_score"].quantile(scale_q)
    lo = rolled["avg_score"].quantile(kill_q)
    scale = rolled[rolled["avg_score"] >= hi].reset_index(drop=True)
    kill = rolled[rolled["avg_score"] <= lo].sort_values("avg_score").reset_index(drop=True)
    return scale, kill


def analyze(df: pd.DataFrame) -> dict:
    """One call -> everything the brief needs: scored clips + rollups + verdicts."""
    scored = compute_scores(df)
    combo = rollup(scored, ACTION_DIMS)
    scale, kill = scale_and_kill(combo)
    return {
        "scored": scored,
        "by_combo": combo,
        "by_creator": rollup(scored, ["source_creator"]),
        "by_format": rollup(scored, ["fmt"]),
        "by_hook": rollup(scored, ["hook_type"]),
        "by_length": rollup(scored, ["length_bucket"]),
        "by_platform": rollup(scored, ["platform"]),
        "scale": scale,
        "kill": kill,
    }

File: rust/src/scoring.rs
//! Scoring engine — the deterministic core. Mirrors src/ycp/scoring.py: per-clip 0-100
//! virality score, then rollups by dimension + scale/kill verdicts. Pure, reproducible.
use std::collections::BTreeMap;

use chrono::{DateTime, Timelike};
use chrono_tz::Tz;

use crate::db::ClipRow;

#[derive(Clone, Debug, Default)]
pub struct Weights {
    pub views_7d: f64,
    pub retention: f64,
    pub revenue_per_1k: f64,
}

pub struct ScoreCfg {
    pub weights: Weights,
    pub length_buckets: Vec<i64>,
    pub min_sample: i64,
    pub scale_q: f64,
    pub kill_q: f64,
    pub timezone: String,
}

impl ScoreCfg {
    pub fn from_settings(s: &serde_yaml::Value) -> Self {
        let sc = &s["scoring"];
        let w = &sc["weights"];
        let f = |v: &serde_yaml::Value, k: &str, d: f64| v[k].as_f64().unwrap_or(d);
        let buckets = sc["length_buckets"]
            .as_sequence()
            .map(|seq| seq.iter().filter_map(|x| x.as_i64()).collect())
            .unwrap_or_else(|| vec![0, 15, 25, 35, 45, 60]);
        ScoreCfg {
            weights: Weights {
                views_7d: f(w, "views_7d", 0.5),
                retention: f(w, "retention", 0.3),
                revenue_per_1k: f(w, "revenue_per_1k", 0.2),
            },
            length_buckets: buckets,
            min_sample: sc["min_sample"].as_i64().unwrap_or(4),
            scale_q: f(sc, "scale_quantile", 0.8),
            kill_q: f(sc, "kill_quantile", 0.2),
            timezone: s["distribution"]["postiz"]["timezone"]
                .as_str()
                .unwrap_or("UTC")
                .to_string(),
        }
    }
}

#[derive(Clone, Debug)]
pub struct Scored {
    #[allow(dead_code)] // parity with the Python dict shape; read at cutover
    pub clip_id: String,
    pub source_creator: String,
    pub fmt: String,
    pub hook_type: String,
    pub platform: String,
    pub length_bucket: String,
    pub post_hour: Option<i64>,
    pub views: f64,
    pub virality_score: f64,
    pub ad_revenue: f64,
    pub swipe_away_pct: Option<f64>,
}

#[derive(Clone, Debug)]
pub struct Rollup {
    pub key: String,
    pub n: i64,
    pub avg_score: f64,
    pub avg_views: f64,
    pub total_revenue: f64,
}

pub struct Analysis {
    pub scored: Vec<Scored>,
    pub by_combo: Vec<Rollup>,
    pub by_creator: Vec<Rollup>,
    pub by_format: Vec<Rollup>,
    pub by_hook: Vec<Rollup>,
    pub by_length: Vec<Rollup>,
    pub by_platform: Vec<Rollup>,
    pub by_hour: Vec<Rollup>,
    pub scale: Vec<Rollup>,
    pub kill: Vec<Rollup>,
}

fn length_bucket(seconds: Option<i64>, buckets: &[i64]) -> String {
    let s = match seconds {
        Some(v) => v,
        None => return "unknown".into(),
    };
    for w in buckets.windows(2) {
        if w[0] <= s && s < w[1] {
            return format!("{}-{}s", w[0], w[1]);
        }
    }
    format!("{}s+", buckets.last().copied().unwrap_or(0))
}

fn post_hour(posted_at: &Option<String>, tz: &str) -> Option<i64> {
    let raw = posted_at.as_ref()?;
    let dt = DateTime::parse_from_rfc3339(raw).ok()?;
    let zone: Tz = tz.parse().unwrap_or(chrono_tz::UTC);
    Some(dt.with_timezone(&zone).hour() as i64)
}

/// Scale a slice to 0..1. Constant or empty → 0.5 (neutral, no signal). Mirrors _minmax.
fn minmax(vals: &[f64]) -> Vec<f64> {
    if vals.is_empty() {
        return vec![];
    }
    let lo = vals.iter().cloned().fold(f64::INFINITY, f64::min);
    let hi = vals.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    if !lo.is_finite() || !hi.is_finite() || hi == lo {
        return vec![0.5; vals.len()];
    }
    vals.iter().map(|v| (v - lo) / (hi - lo)).collect()
}

pub fn compute_scores(rows: &[ClipRow], cfg: &ScoreCfg) -> Vec<Scored> {
    if rows.is_empty() {
        return vec![];
    }
    let views: Vec<f64> = rows
        .iter()
        .map(|r| (r.views.max(0) as f64).ln_1p())
        .collect();
    let ret: Vec<f64> = rows
        .iter()
        .map(|r| r.retention_pct.unwrap_or(0.0).max(0.0))
        .collect();
    let rev1k: Vec<f64> = rows
        .iter()
        .map(|r| {
            let v = r.views.max(0) as f64;
            if v > 0.0 {
                r.ad_revenue / (v / 1000.0)
            } else {
                0.0
            }
        })
        .collect();
    let (nv, nr, nrev) = (minmax(&views), minmax(&ret), minmax(&rev1k));
    let w = &cfg.weights;
    rows.iter()
        .enumerate()
        .map(|(i, r)| {
            let score = w.views_7d * nv[i] + w.retention * nr[i] + w.revenue_per_1k * nrev[i];
            Scored {
                clip_id: r.clip_id.clone(),
                source_creator: r.source_creator.clone().unwrap_or_default(),
                fmt: r.fmt.clone().unwrap_or_default(),
                hook_type: r.hook_type.clone().unwrap_or_default(),
                platform: r.platform.clone().unwrap_or_default(),
                length_bucket: length_bucket(r.length_sec, &cfg.length_buckets),
                post_hour: post_hour(&r.posted_at, &cfg.timezone),
                views: r.views as f64,
                // Python `(score*100).round(1)` — round-half-to-even via util::round_to.
                virality_score: crate::util::round_to(score * 100.0, 1),
                ad_revenue: r.ad_revenue,
                swipe_away_pct: r.swipe_away_pct,
            }
        })
        .collect()
}

/// Group by a key, keep combos with >= min_sample clips, sort by avg_score desc.
pub fn rollup<F: Fn(&Scored) -> String>(scored: &[Scored], key: F, min_sample: i64) -> Vec<Rollup> {
    let mut groups: BTreeMap<String, (i64, f64, f64, f64)> = BTreeMap::new();
    for s in scored {
        let e = groups.entry(key(s)).or_insert((0, 0.0, 0.0, 0.0));
        e.0 += 1;
        e.1 += s.virality_score;
        e.2 += s.views;
        e.3 += s.ad_revenue;
    }
    let mut out: Vec<Rollup> = groups
        .into_iter()
        .filter(|(_, (n, ..))| *n >= min_sample)
        .map(|(key, (n, sc, vw, rev))| Rollup {
            // Python pandas `.round(n)` is round-half-to-even; util::round_to reproduces it.
            // Rust's bare `.round()` is round-half-away → diverges on exact .5 ties (e.g. an
            // avg_views of 0.5 → 0 in Python but 1 with bare round()). Surfaced by autopilot.
            key,
            n,
            avg_score: crate::util::round_to(sc / n as f64, 1),
            avg_views: crate::util::round_to(vw / n as f64, 0),
            total_revenue: crate::util::round_to(rev, 2),
        })
        .collect();
    out.sort_by(|a, b| b.avg_score.partial_cmp(&a.avg_score).unwrap());
    out
}

/// Linear-interpolation quantile (matches pandas default).
fn quantile(sorted: &[f64], q: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    let pos = q * (sorted.len() - 1) as f64;
    let lo = pos.floor() as usize;
    let hi = pos.ceil() as usize;
    if lo == hi {
        sorted[lo]
    } else {
        sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo as f64)
    }
}

pub fn scale_and_kill(rolled: &[Rollup], scale_q: f64, kill_q: f64) -> (Vec<Rollup>, Vec<Rollup>) {
    if rolled.is_empty() {
        return (vec![], vec![]);
    }
    let mut scores: Vec<f64> = rolled.iter().map(|r| r.avg_score).collect();
    scores.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let hi = quantile(&scores, scale_q);
    let lo = quantile(&scores, kill_q);
    let scale: Vec<Rollup> = rolled
        .iter()
        .filter(|r| r.avg_score >= hi)
        .cloned()
        .collect();
    let mut kill: Vec<Rollup> = rolled
        .iter()
        .filter(|r| r.avg_score <= lo)
        .cloned()
        .collect();
    kill.sort_by(|a, b| a.avg_score.partial_cmp(&b.avg_score).unwrap());
    (scale, kill)
}

pub fn analyze(rows: &[ClipRow], cfg: &ScoreCfg) -> Analysis {
    let scored = compute_scores(rows, cfg);
    let by_combo = rollup(
        &scored,
        |s| format!("{} · {} · {}", s.source_creator, s.fmt, s.hook_type),
        cfg.min_sample,
    );
    let (scale, kill) = scale_and_kill(&by_combo, cfg.scale_q, cfg.kill_q);
    Analysis {
        by_creator: rollup(&scored, |s| s.source_creator.clone(), cfg.min_sample),
        by_format: rollup(&scored, |s| s.fmt.clone(), cfg.min_sample),
        by_hook: rollup(&scored, |s| s.hook_type.clone(), cfg.min_sample),
        by_length: rollup(&scored, |s| s.length_bucket.clone(), cfg.min_sample),
        by_platform: rollup(&scored, |s| s.platform.clone(), cfg.min_sample),
        by_hour: rollup(
            &scored
                .iter()
                .filter(|s| s.post_hour.is_some())
                .cloned()
                .collect::<Vec<_>>(),
            |s| s.post_hour.unwrap().to_string(),
            cfg.min_sample,
        ),
        by_combo,
        scale,
        kill,
        scored,
    }
}
